esphome/components/substitutions/__init__.py

```diff
@@ -65,25 +97,32 @@ def _expand_jinja(value, orig_value, path, jinja, ignore_missing):
                 f"\nSee {'->'.join(str(x) for x in path)}",
                 path,
             )
+        # If the original, unexpanded string, contained document metadata (ESPHomeDatabase),
+        # assign this same document metadata to the resulting value.
+        if isinstance(orig_value, ESPHomeDataBase):
+            value = _restore_data_base(value, orig_value)
+
     return value
 
 
-def _expand_substitutions(substitutions, value, path, jinja, ignore_missing):
+def _expand_substitutions(
+    substitutions: dict, value: str, path, jinja: Jinja, ignore_missing: bool
+) -> Any:
     if "$" not in value:
         return value
 
-    orig_value = value
+    orig_value: str = value
 
-    i = 0
+    i: int = 0
```

Comment:
```suggestion
    i = 0
```

This can be worked out automatically since its assigned immediately

Reply:
ok, thanks.